lib/firmware_model.dart

```diff
@@ -50,6 +50,8 @@ class FirmwareModel extends SafeChangeNotifier {
   Future<void> install(FwupdDevice device, FwupdRelease release) async {
     try {
       await _service.install(device, release);
+      await refresh();
+      return _updateState();
```

Comment:
using `ChangeNotifierProxyProvider` should be enough because the `_updateState()` gets automatically called when `FwupdNotifier` sees a device update

Reply:
ok, that means calling `_updateState()` is unnecessary, but we still need to call `refresh()` to update the device list

Reply:
fwupd emits a "device changed" event whenever anything changes. whenever this happens, the app should output something like this:
```
flutter: DEBUG fwupd_notifier: updated FwupdDevice(checksum: null, created: 2022-09-19 08:09:08.000Z, deviceId: 08d460be0f1f9f128413f816022a6439e0078018, flags: {FwupdDeviceFlag.requireAc, FwupdDeviceFlag.supported, FwupdDeviceFlag.registered, FwupdDeviceFlag.canVerify, FwupdDeviceFlag.canVerifyImage}, guid: [b585990a-003e-5270-89d5-3705a17f9a43], icon: [preferences-desktop-keyboard], modified: 2022-09-22 09:32:14.000Z, name: 'Integrated Webcam™', parentDeviceId: null, plugin: test, protocol: com.acme.test, summary: Fake webcam, updateState: FwupdUpdateState.success, vendor: ACME Corp., vendorId: USB:0x046D, version: 1.2.4, versionBootloader: 0.1.2, versionFormat: FwupdVersionFormat.triplet, versionLowest: 1.2.0)
```
`FirmwareModel` listens to these updates and calls `_updateState` so it should be fine (tm) as long as `DeviceModel` watches/proxies/listens to `FirmwareModel` instead of reading :)

Reply:
> fwupd emits a "device changed" event whenever anything changes.

i mean any device attributes change including status and version changes which happen during installation

Reply:
hm, all of this makes sense to me - however if I remove the `refresh()` call the `DevicePage` doesn't update:

[Screencast from 22-09-22 11:41:11.webm](https://user-images.githubusercontent.com/113362648/191714654-48b6f733-dfb9-43eb-b60d-9fd97ba5c94f.webm)

Reply:
strange! works for me with the `ChangeNotifierProxyProvider` change alone. how does the debug output look?

Reply:
```
Running "flutter pub get" in firmware-updater...
Launching lib/main.dart on Linux in debug mode...
Connecting to VM Service at ws://127.0.0.1:33477/lFjjlwcHcys=/ws
flutter: DEBUG fwupd_notifier: 12 devices
flutter: DEBUG fwupd_service: install FwupdRelease(appstreamId: fakedevice.firmware, checksum: fc0aabcf98bf3546c91270f2941f0acd0395dd79, created: 2017-06-15 04:00:00.000Z, description: '<p>Fixes another bug with the flux capacitor to prevent time going backwards.</p>', filename: 2b8546ba805ad10bf8a2e5ad539d53f303812ba5, homepage: http://www.acme.com/, license: GPL-2.0+, locations: [./fakedevice124.cab], name: 'FakeDevice Firmware', protocol: null, remoteId: fwupd-tests, size: 17, flags: {FwupdReleaseFlag.isUpgrade}, urgency: FwupdReleaseUrgency.medium, uri: ./fakedevice124.cab, vendor: 'ACME Corp', version: '1.2.4')
flutter: DEBUG fwupd_service: on FwupdDevice(checksum: null, created: 2022-09-22 08:21:24.000Z, deviceId: 08d460be0f1f9f128413f816022a6439e0078018, flags: {FwupdDeviceFlag.updatable, FwupdDeviceFlag.requireAc, FwupdDeviceFlag.supported, FwupdDeviceFlag.registered, FwupdDeviceFlag.canVerify, FwupdDeviceFlag.canVerifyImage}, guid: [b585990a-003e-5270-89d5-3705a17f9a43], icon: [preferences-desktop-keyboard], modified: 2022-09-22 09:58:40.000Z, name: 'Integrated Webcam™', parentDeviceId: null, plugin: test, protocol: com.acme.test, summary: Fake webcam, updateState: FwupdUpdateState.success, vendor: ACME Corp., vendorId: USB:0x046D, version: 1.2.3, versionBootloader: 0.1.2, versionFormat: FwupdVersionFormat.triplet, versionLowest: 1.2.0)
flutter: DEBUG fwupd_notifier: FwupdStatus.decompressing
flutter: DEBUG fwupd_notifier: FwupdStatus.waitingForAuth
flutter: DEBUG fwupd_notifier: FwupdStatus.waitingForAuth
flutter: DEBUG fwupd_notifier: FwupdStatus.idle
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceWrite
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceWrite
flutter: DEBUG fwupd_notifier: 2%
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceWrite
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceWrite
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceRestart
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceRestart
flutter: DEBUG fwupd_notifier: 96%
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceBusy
flutter: DEBUG fwupd_notifier: 98%
flutter: DEBUG fwupd_notifier: FwupdStatus.unknown
flutter: DEBUG fwupd_notifier: 100%
Lost connection to device.
Exited (sigterm)
```

Reply:
hmm, not a single device update? do integration tests pass?

```
Restarted application in 186ms.
flutter: DEBUG fwupd_notifier: 9 devices
flutter: DEBUG fwupd_service: install FwupdRelease(appstreamId: fakedevice.firmware, checksum: fc0aabcf98bf3546c91270f2941f0acd0395dd79, created: 2017-06-15 04:00:00.000Z, description: '<p>Fixes another bug with the flux capacitor to prevent time going backwards.</p>', filename: 2b8546ba805ad10bf8a2e5ad539d53f303812ba5, homepage: http://www.acme.com/, license: GPL-2.0+, locations: [./fakedevice124.cab], name: 'FakeDevice Firmware', protocol: null, remoteId: fwupd-tests, size: 17, flags: {FwupdReleaseFlag.isUpgrade}, urgency: FwupdReleaseUrgency.medium, uri: ./fakedevice124.cab, vendor: 'ACME Corp', version: '1.2.4')
flutter: DEBUG fwupd_service: on FwupdDevice(checksum: null, created: 2022-09-19 08:09:08.000Z, deviceId: 08d460be0f1f9f128413f816022a6439e0078018, flags: {FwupdDeviceFlag.updatable, FwupdDeviceFlag.requireAc, FwupdDeviceFlag.supported, FwupdDeviceFlag.registered, FwupdDeviceFlag.canVerify, FwupdDeviceFlag.canVerifyImage}, guid: [b585990a-003e-5270-89d5-3705a17f9a43], icon: [preferences-desktop-keyboard], modified: 2022-09-22 10:03:15.000Z, name: 'Integrated Webcam™', parentDeviceId: null, plugin: test, protocol: com.acme.test, summary: Fake webcam, updateState: FwupdUpdateState.success, vendor: ACME Corp., vendorId: USB:0x046D, version: 1.2.3, versionBootloader: 0.1.2, versionFormat: FwupdVersionFormat.triplet, versionLowest: 1.2.0)
flutter: DEBUG fwupd_notifier: FwupdStatus.waitingForAuth
flutter: DEBUG fwupd_notifier: FwupdStatus.waitingForAuth
flutter: DEBUG fwupd_notifier: 0%
flutter: DEBUG fwupd_notifier: updated FwupdDevice(checksum: null, created: 2022-09-19 08:09:08.000Z, deviceId: 08d460be0f1f9f128413f816022a6439e0078018, flags: {FwupdDeviceFlag.requireAc, FwupdDeviceFlag.supported, FwupdDeviceFlag.registered, FwupdDeviceFlag.canVerify, FwupdDeviceFlag.canVerifyImage}, guid: [b585990a-003e-5270-89d5-3705a17f9a43], icon: [preferences-desktop-keyboard], modified: 2022-09-22 10:03:26.000Z, name: 'Integrated Webcam™', parentDeviceId: null, plugin: test, protocol: com.acme.test, summary: Fake webcam, updateState: FwupdUpdateState.success, vendor: ACME Corp., vendorId: USB:0x046D, version: 1.2.3, versionBootloader: 0.1.2, versionFormat: FwupdVersionFormat.triplet, versionLowest: 1.2.0)
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceBusy
flutter: DEBUG fwupd_notifier: updated FwupdDevice(checksum: null, created: 2022-09-19 08:09:08.000Z, deviceId: 08d460be0f1f9f128413f816022a6439e0078018, flags: {FwupdDeviceFlag.requireAc, FwupdDeviceFlag.supported, FwupdDeviceFlag.registered, FwupdDeviceFlag.canVerify, FwupdDeviceFlag.canVerifyImage, FwupdDeviceFlag.updatableHidden}, guid: [b585990a-003e-5270-89d5-3705a17f9a43], icon: [preferences-desktop-keyboard], modified: 2022-09-22 10:03:26.000Z, name: 'Integrated Webcam™', parentDeviceId: null, plugin: test, protocol: com.acme.test, summary: Fake webcam, updateState: FwupdUpdateState.success, vendor: ACME Corp., vendorId: USB:0x046D, version: 1.2.3, versionBootloader: 0.1.2, versionFormat: FwupdVersionFormat.triplet, versionLowest: 1.2.0)
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceWrite
flutter: DEBUG fwupd_notifier: 1%
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceWrite
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceWrite
flutter: DEBUG fwupd_notifier: 2%
flutter: DEBUG fwupd_notifier: FwupdStatus.decompressing
flutter: DEBUG fwupd_notifier: updated FwupdDevice(checksum: null, created: 2022-09-19 08:09:08.000Z, deviceId: 08d460be0f1f9f128413f816022a6439e0078018, flags: {FwupdDeviceFlag.requireAc, FwupdDeviceFlag.supported, FwupdDeviceFlag.registered, FwupdDeviceFlag.canVerify, FwupdDeviceFlag.canVerifyImage, FwupdDeviceFlag.updatableHidden}, guid: [b585990a-003e-5270-89d5-3705a17f9a43], icon: [preferences-desktop-keyboard], modified: 2022-09-22 10:03:26.000Z, name: 'Integrated Webcam™', parentDeviceId: null, plugin: test, protocol: com.acme.test, summary: Fake webcam, updateState: FwupdUpdateState.success, vendor: ACME Corp., vendorId: USB:0x046D, version: 1.2.4, versionBootloader: 0.1.2, versionFormat: FwupdVersionFormat.triplet, versionLowest: 1.2.0)
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceVerify
flutter: DEBUG fwupd_notifier: updated FwupdDevice(checksum: null, created: 2022-09-19 08:09:08.000Z, deviceId: 08d460be0f1f9f128413f816022a6439e0078018, flags: {FwupdDeviceFlag.requireAc, FwupdDeviceFlag.supported, FwupdDeviceFlag.registered, FwupdDeviceFlag.canVerify, FwupdDeviceFlag.canVerifyImage}, guid: [b585990a-003e-5270-89d5-3705a17f9a43], icon: [preferences-desktop-keyboard], modified: 2022-09-22 10:03:26.000Z, name: 'Integrated Webcam™', parentDeviceId: null, plugin: test, protocol: com.acme.test, summary: Fake webcam, updateState: FwupdUpdateState.success, vendor: ACME Corp., vendorId: USB:0x046D, version: 1.2.4, versionBootloader: 0.1.2, versionFormat: FwupdVersionFormat.triplet, versionLowest: 1.2.0)
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceRestart
flutter: DEBUG fwupd_notifier: updated FwupdDevice(checksum: null, created: 2022-09-19 08:09:08.000Z, deviceId: 08d460be0f1f9f128413f816022a6439e0078018, flags: {FwupdDeviceFlag.updatable, FwupdDeviceFlag.requireAc, FwupdDeviceFlag.supported, FwupdDeviceFlag.registered, FwupdDeviceFlag.canVerify, FwupdDeviceFlag.canVerifyImage}, guid: [b585990a-003e-5270-89d5-3705a17f9a43], icon: [preferences-desktop-keyboard], modified: 2022-09-22 10:03:26.000Z, name: 'Integrated Webcam™', parentDeviceId: null, plugin: test, protocol: com.acme.test, summary: Fake webcam, updateState: FwupdUpdateState.success, vendor: ACME Corp., vendorId: USB:0x046D, version: 1.2.4, versionBootloader: 0.1.2, versionFormat: FwupdVersionFormat.triplet, versionLowest: 1.2.0)
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceRestart
flutter: DEBUG fwupd_notifier: 95%
flutter: DEBUG fwupd_notifier: FwupdStatus.deviceBusy
flutter: DEBUG fwupd_notifier: 99%
flutter: DEBUG fwupd_notifier: 100%
flutter: DEBUG fwupd_notifier: 100%
flutter: DEBUG fwupd_notifier: FwupdStatus.idle
flutter: DEBUG fwupd_notifier: FwupdStatus.idle
```

Reply:
Yes, all tests pass

Reply:
have you ever seen those device update messages while working in the app? 👀

Reply:
I don't have a history of all the debug messages, but I definitely haven't seen them while investigating this bug.

Reply:
I can reproduce the issue on 22.04. :+1: There are no device updates during the installation. Plugging the power cable in/out does trigger an update, though.

Reply:
for me the problem only happens with the fake webcam test device on 22.04 (fwupd 1.7.5) - all real device send the expected `DeviceChange` signal on DBus. In any case the problem shouldn't be solved by forcing another update of the device list - I've removed those lines.